pub/helpers/bin/infer_sdk_versions.dart

```diff
@@ -0,0 +1,162 @@
+  return {
+    'dart': dartConstraint,
+    'flutter': flutterConstraint,
+  };
+}
+
+Future<void> main(List<String> args) async {
+  try {
+    final argResults = parseArgs(args);
+    final flutterReleases = await retrieveFlutterReleases(
+        argResults['flutter-releases-url'] ?? flutterReleasesUrl);
+
+    final pubspecPath = p.join(argResults['directory'], 'pubspec.yaml');
+    final pubspec = loadYaml(File(pubspecPath).readAsStringSync(),
+        sourceUrl: Uri.file(pubspecPath));
+
+    final bestFlutterRelease =
+        inferBestFlutterRelease(parseSdkConstraints(pubspec), flutterReleases);
+    if (bestFlutterRelease == null) {
+      fail(
+        'No flutter release matching sdk constraints.',
+      );
+    }
+    stdout.writeln(JsonEncoder.withIndent('  ').convert({
+      'flutter': bestFlutterRelease.flutterVersion.toString(),
+      'dart': bestFlutterRelease.dartVersion.toString(),
+      'channel': {
+        Channel.stable: 'stable',
+        Channel.beta: 'beta',
+        Channel.dev: 'dev'
+      }[bestFlutterRelease.channel],
+    }));
+  } on FormatException catch (e) {
+    fail(e.message);
+  } finally {
+    client.close();
+  }
+}
+
+String get flutterReleasesUrl =>
+    'https://storage.googleapis.com/flutter_infra_release/releases/releases_linux.json';
+
+// Retrieves all released versions of Flutter.
+Future<List<FlutterRelease>> retrieveFlutterReleases(String url) async {
+  final response = await client.get(Uri.parse(flutterReleasesUrl));
+  final decoded = jsonDecode(response.body);
+  if (decoded is! Map) throw FormatException('Bad response - should be a Map');
+  final releases = decoded['releases'];
+  if (releases is! List)
+    throw FormatException('Bad response - releases should be a list.');
+  final result = <FlutterRelease>[];
+  for (final release in releases) {
+    final channel = {
+      'beta': Channel.beta,
+      'stable': Channel.stable,
+      'dev': Channel.dev
+    }[release['channel']];
+    if (channel == null) throw FormatException('Release with bad channel');
+    final dartVersion = release['dart_sdk_version'];
+    // Some releases don't have an associated dart version, ignore.
+    if (dartVersion is! String) continue;
+    final flutterVersion = release['version'];
+    if (flutterVersion is! String) throw FormatException('Not a string');
+    result.add(FlutterRelease(
+      flutterVersion: Version.parse(flutterVersion),
+      dartVersion: Version.parse(dartVersion.split(' ').first),
+      channel: channel,
+    ));
+  }
+  return result
+      // Sort releases by channel and version.
+      .sorted((a, b) {
+        final compareChannels = b.channel.index - a.channel.index;
+        if (compareChannels != 0) return compareChannels;
+        return a.flutterVersion.compareTo(b.flutterVersion);
+      })
+      // Newest first.
+      .reversed
+      .toList();
+}
+
+/// The "best" Flutter release for a given set of constraints is the first one
```

Comment:
does "first" mean: oldest or newest Flutter version?

Reply:
It is the first in the list passed to this function.

I tried to rephrase a bit.